### PR TITLE
Add visual tests for react-native-xaml

### LINKED ISSUE #9069: Add tests for VM that uses native layout
We don't have tests for View Managers that use RequiresNativeLayout, and layout is a fragile area, so we should add some tests. 

One idea would be to add a page in e2e-test-app that uses react-native-xaml, and use the snapshots infra to dump the position/sizes of elements to validate they are as expected.

### PR DESCRIPTION
**Pull request by asklar**
Fixes #9069 to validate visual layout is preserved for react-native-xaml

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9083)

### CODE REVIEW

**Comment by NickGerleman** [comment]
Does playground need to be fixed to support this page now?

File: packages/@react-native-windows/tester/package.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "@react-native-picker/picker": "2.2.0",
     "react": "17.0.2",
     "react-native": "0.0.0-20211021-2008-eccbf9b5c",
-    "react-native-windows": "^0.0.0-canary.414"
+    "react-native-windows": "^0.0.0-canary.414",
+    "react-native-xaml": "^0.0.50"
```
Comment (NickGerleman) [inline packages/@react-native-windows/tester/package.json:21]: Are any APIs changing? If so should we pin the version here?

**Comment by NickGerleman** [comment]
What's the motivation of having the tests here instead of in the react-native-xaml repo? Is it to try to catch specific platform changes?

File: packages/@react-native-windows/tester/package.json (hunk 1 above)
Comment (NickGerleman) [reply packages/@react-native-windows/tester/package.json:21]: Esp if we are snapshotting the output of its components.

**Comment by asklar** [comment]
> What's the motivation of having the tests here instead of in the react-native-xaml repo? Is it to try to catch specific platform changes?

yes, esp want to protect against regressing layout changes (and add coverage for things like RequiresNativeLayout), before it reaches consuming repos like react-native-xaml.

File: packages/@react-native-windows/tester/package.json (hunk 1 above)
Comment (asklar) [reply packages/@react-native-windows/tester/package.json:21]: apis are coming from xaml so they should be stable

File: packages/e2e-test-app/test/XAMLTest.test.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    const tappableTextBlock = await app.findElementByTestID(
+      'tappableTextBlock',
+    );
+    await tappableTextBlock.click();
+    const dump = await dumpVisualTree('MenuFlyout');
+    expect(dump).toMatchSnapshot('ReactNativeXAML-MenuFlyout');
+    const menuOption2 = await app.findElementByTestID('menuOption2');
+    await menuOption2.click();
+    const text = await tappableTextBlock.getText();
+    expect(text).toBe('menu option 2 clicked');
```
Comment (NickGerleman) [inline packages/e2e-test-app/test/XAMLTest.test.ts:31]: There is a race here. We sent a click via UIA, but the text is updated asynchronously. You can use `app.WaitUntil` to poll for the condition to be met.

Note that the above `click`, etc are fine. WinAppDriver will retry finding an alement until it is available. A first valid value will be returned after performing an action though.

File: packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js [hunk 3]
```diff
@@ -51,6 +51,11 @@ const Components: Array<RNTesterModuleInfo> = [
     category: 'UI',
     module: require('../examples-win/Glyph/GlyphExample'),
   },
+  {
+    key: 'XAML',
+    category: 'UI',
+    module: require('../examples-win/XAML/XAMLExample'),
+  },
```
Comment (NickGerleman) [inline packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js:58]: Keep alphabetized

**Comment by NickGerleman** [comment]
I overall think it's okay to add this, but we probably want to think a bit more generically about how we can test external module compat before adding anything more. I don't think adding them to RNTester will scale.

Not sure what level of coverage react-native-xaml has in its own repo, but most of the test infra behind `e2e-test-app` can now be used in other repos. That doesn't help catch RNW changes in PR though, so I do see value in this change.

Might be nice to isolate tests to the specific APIs used by react-native-xaml. I added a frozen-in-time version of Picker as a quick way to get some coverage, and replace our own Picker usages, but I expect `react-native-xaml` usage may be more nuanced. I'm fine seeing what happens though, and if any issues pop up in the future.

File: packages/@react-native-windows/tester/package.json (hunk 1 above)
Comment (NickGerleman) [reply packages/@react-native-windows/tester/package.json:21]: But are `react-native-xaml` projections stable? E.g. I remember a change from strings to enums before, that would be breaking.

File: packages/@react-native-windows/tester/src/js/examples-win/XAML/XAMLExample.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        onClosed={() => setVisible(Visibility.Collapsed)}
+      />
+    </View>
+  );
+}
+
+export const displayName = (_undefined?: string) => {};
+export const title = 'XAML';
+export const category = 'UI';
+export const description = 'Usage of React Native XAML controls';
```
Comment (NickGerleman) [inline packages/@react-native-windows/tester/src/js/examples-win/XAML/XAMLExample.tsx:114]: nit: specify `react-native-xaml`

File: packages/e2e-test-app/test/XAMLTest.test.ts (hunk 2 above)
Comment (asklar) [reply packages/e2e-test-app/test/XAMLTest.test.ts:31]: good catch.

one thing I'm seeing (unrelated) is that the dump for the menu isn't generated/collected in CI... any idea what's missing?

Comment (NickGerleman) [reply packages/e2e-test-app/test/XAMLTest.test.ts:31]: It looks like something is crashing. Reminder for me to figure out why dumps aren't being collected...

Usually when I've seen CI crashes that do not happen locally, it is from using APIs that are unsupported on Server 2019 (Windows 17763). Something to check for here.

File: packages/e2e-test-app/test/XAMLTest.test.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+beforeAll(async () => {
+  await goToComponentExample('XAML');
+});
+
+describe('XamlTest', () => {
+  test('Layout', async () => {
+    const dump = await dumpVisualTree('ReactNativeXAMLRoot');
+    expect(dump).toMatchSnapshot();
+  });
+  test('MenuFlyoutIsOpen', async () => {
```
Comment (NickGerleman) [inline packages/e2e-test-app/test/XAMLTest.test.ts:21]: You can put spaces in jest tests titles if you want to

File: packages/e2e-test-app/test/XAMLTest.test.ts (hunk 2 above)
Comment (NickGerleman) [reply packages/e2e-test-app/test/XAMLTest.test.ts:31]: It's also possible the e2e-test-app output is bad... The error messages we have right now leave a bit to be desired.

I think you might also want to poll in these two lines, to ensure the flyout is open before trying to dump it.

```
await tappableTextBlock.click();
const dump = await dumpVisualTree('MenuFlyout');
```

**Review by NickGerleman — APPROVED** [review]
Looks good. Please hook up any modules required by `@react-native-windows/tester` to the playground project before merging.

File: packages/e2e-test-app/test/__snapshots__/XAMLTest.test.ts.snap [hunk 6]
```diff
@@ -0,0 +1,780 @@
...
+      ],
+    },
+  ],
+}
+`;
+
+exports[`ReactNativeXAML-MenuFlyout`] = `
+Object {
+}
+`;
```
Comment (NickGerleman) [inline packages/e2e-test-app/test/__snapshots__/XAMLTest.test.ts.snap:780]: I think this bit is causing jest to fail

Comment (asklar) [reply packages/e2e-test-app/test/__snapshots__/XAMLTest.test.ts.snap]: yeah looks like snapshot names must end with a number (!!)

**Review by dannyvv — APPROVED** [review]

File: packages/e2e-test-app/test/XAMLTest.test.ts (hunk 2 above)
Comment (asklar) [reply packages/e2e-test-app/test/XAMLTest.test.ts]: I figured this one out - the menuFlyout doesn't actually end up in the visual tree. XAML will create a MenuFlyoutPresenter (which doesn't carry the automation ID from the MenuFlyout) and put that in a separate visual root separate from the App's. The XAML VisualTreeHelper APIs don't seem to be able to reason over children of popups, so we need some special casing, see latest commit.

**Comment by asklar** [comment]
@NickGerleman I tried adding rnxaml to playground, but we don't yet support it with winui3, and since playground is built for winui3 as well, this is a problem; I tried only including the project reference and package provider stuff when not using winui3, but the playground sln file still needs to reference it, and since we build form the sln file in CI, we still try to build it in the winui3 config, which fails.
Going to take that as a follow up item since I don't want to block adding validation for the mainline scenario; we are going to have updated references to winappsdk soon, so the existing winui3 logic will change a lot, meaning it doesn't make much sense to target the old winui3 preview 4 from before winappsdk now anyway.

**Comment by NickGerleman** [comment]
Sounds reasonable.

**Pull request merged**